### ISSUE TITLE
x/wasm hooks

### ISSUE THREAD

**Issue by jhernandezb** [CosmWasm/wasmd#1683]
For a couple of months I've been thinking about a cosmwasm system where there can be namespaces: ability to upload codes to a namespace where they can interact each other but also open permissions to other namespaces you can think of like linux namespaces and a firewall.

First I thought just playing around with a fork but then I think that this could also be achieved if another module had access to hooks like
```
BeforeInstantiate()
BeforeExecute()
BeforeStore()
BeforeMigrate()
```
then an`x/namespace` module can just subscribe to hooks and apply whatever logic.

This is just an idea but not sure if others would find this interesting our useful

**Comment by alpe** [comment]
Thanks for bringing this topic up. I am happy to see new ideas and use cases rising.
We could replace the keeper instance with an interface in [msg_server](https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/msg_server.go#L21) so that you can play around without the need for a fork for now.

**Comment by jhernandezb** [comment]
> Thanks for bringing this topic up. I am happy to see new ideas and use cases rising. We could replace the keeper instance with an interface in [msg_server](https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/msg_server.go#L21) so that you can play around without the need for a fork for now.

Are intra contract calls routed through msg_server? If so that would help  me with exactly what I want

**Comment by alpe** [comment]
> Are intra contract calls routed through msg_server? 

yes, you can find all the message handling in https://github.com/CosmWasm/wasmd/blob/v0.44.0/x/wasm/keeper/handler_plugin.go#L54

The wasm keeper has private functions to not mess with the authorization system but we could provide some adapter here to help you out
